src/renderer/object/instanced_model.rs

```diff
@@ -49,53 +52,173 @@ impl<M: ForwardMaterial> InstancedModel<M> {
             instance_buffer1: InstanceBuffer::new(context)?,
             instance_buffer2: InstanceBuffer::new(context)?,
             instance_buffer3: InstanceBuffer::new(context)?,
+            instance_buffer4: InstanceBuffer::new(context)?,
             aabb,
             aabb_local: aabb.clone(),
             transformation: Mat4::identity(),
             transformations: transformations.to_vec(),
+            subtexture: None,
+            subtextures: None,
             material,
         };
-        model.update_transformations(transformations);
-        model.update_aabb();
+        model.set_transformations(transformations);
         Ok(model)
     }
 
     ///
-    /// Updates the transformations applied to each model instance before they are rendered.
-    /// The model is rendered in as many instances as there are transformation matrices.
+    /// Use [set_transformations()](InstancedModel::set_transformations()) instead.
     ///
+    #[deprecated]
     pub fn update_transformations(&mut self, transformations: &[Mat4]) {
+        self.set_transformations(&transformations);
+    }
+
+    ///
+    /// Store the transformations applied to each model instance before they are rendered.
+    /// The model is rendered in as many instances as there are transformation matrices.
+    ///
+    pub fn set_transformations(&mut self, transformations: &[Mat4]) {
         self.transformations = transformations.to_vec();
-        self.instance_count = transformations.len() as u32;
+        self.update_buffers();
+    }
+
+    ///
+    /// Store the transformations applied to each model instance before they are rendered.
+    /// The model is rendered in as many instances as there are transformation matrices.
+    /// Must (eventually) call [update_buffers()](InstancedModel::update_buffers()) afterward.
+    ///
+    pub fn add_transformation(&mut self, transformation: &Mat4) {
+        self.transformations.push(*transformation);
+    }
+
+    ///
+    /// Clear transformations array.
+    ///
+    pub fn clear_transformations(&mut self) {
+        self.transformations = Vec::new();
+        self.update_buffers();
+    }
+
+    ///
+    /// Update instance transform and uv buffers and aabb.
+    ///
+    pub fn update_buffers(&mut self) {
```

Comment:
Should not be public

Reply:
Public for use after a batch of add_transformation/add_subtexture as to not be rebuilding buffers after every single add. Individual for instanced sprite handling. If anything it should be providing an index back for updating later. Does need some massaging.

Reply:
I think it makes it a bit too complicated. If we only provide a `set_instances` function that takes an array of instance definitions (transformation + texture transform), then one has to make that array outside of `three-d` but that should be easy enough, and then no one will forget to call `update_buffers`.